File: packages/domain/src/ai.ts
import type { AIProviderId } from "./enums.js";
import type { ConversationId, NoteId, NotebookId, TagId } from "./ids.js";

/**
 * Provider-agnostic AI port. V1 has one implementation
 * (OllamaProviderAdapter). The facade layer (`AIServices`) is what UI
 * code talks to — never this port directly.
 */
export interface AIProvider {
  readonly id: AIProviderId;
  readonly capabilities: ReadonlySet<Capability>;
  listModels(): Promise<readonly ModelDescriptor[]>;
  chat(messages: readonly Message[], opts: AIOptions): Promise<AIResponse>;
  streamChat(messages: readonly Message[], opts: AIOptions): AsyncIterable<AIChunk>;
  embed(texts: readonly string[], model: string): Promise<readonly number[][]>;
  ping(): Promise<ProviderHealth>;
}

export type Capability = "STREAMING" | "EMBEDDINGS" | "VISION" | "TOOL_USE";

export interface ModelDescriptor {
  readonly id: string;
  readonly contextTokens: number;
  readonly supports: ReadonlySet<Capability>;
}

export type MessageRole = "system" | "user" | "assistant";

export interface Message {
  readonly role: MessageRole;
  readonly content: string;
}

export interface AIOptions {
  readonly model: string;
  readonly temperature?: number;
  /** Hard ceiling so a runaway model doesn't blow the context budget. */
  readonly maxTokens?: number;
  /** Pass-through for provider-specific stop sequences. */
  readonly stop?: readonly string[];
}

export interface AIResponse {
  readonly content: string;
  readonly model: string;
  readonly promptTokens: number;
  readonly completionTokens: number;
  readonly latencyMs: number;
}

/**
 * Each token chunk from a streaming chat. The pipeline emits structured
 * `citation` and `done` events separately — those are not part of this
 * shape. Pipelines that just want raw token text can collect `delta`s.
 */
export interface AIChunk {
  readonly delta: string;
  /** True on the final chunk; carries the same metrics as AIResponse. */
  readonly done: boolean;
  readonly model?: string;
  readonly promptTokens?: number;
  readonly completionTokens?: number;
}

export interface ProviderHealth {
  readonly ok: boolean;
  readonly message?: string;
  readonly latencyMs?: number;
}

// ---------------------------------------------------------------------------
// AIServices facade types
// ---------------------------------------------------------------------------

export type SummaryLength = "short" | "medium" | "long";

export interface NoteScope {
  readonly notebookIds?: readonly NotebookId[];
  readonly tagIds?: readonly TagId[];
  readonly excludeAi?: boolean;
}

export interface ChatChunk {
  readonly kind: "delta" | "citation" | "done" | "warning";
  /** delta */
  readonly delta?: string;
  /** citation */
  readonly noteId?: NoteId;
  readonly title?: string;
  readonly chunkIndex?: number;
  readonly rank?: number;
  /** done */
  readonly conversationId?: ConversationId;
  readonly tokens?: { readonly prompt: number; readonly completion: number };
  readonly model?: string;
  readonly latencyMs?: number;
  /** warning */
  readonly warning?: string;
}

export interface RelatedNote {
  readonly noteId: NoteId;
  readonly title: string;
  readonly snippet: string;
  /** Cosine distance — smaller is closer. */
  readonly distance: number;
}

export interface Briefing {
  readonly date: string; // YYYY-MM-DD
  readonly markdown: string;
  readonly model: string;
  readonly tokens: { readonly prompt: number; readonly completion: number };
}

export interface Conversation {
  readonly id: ConversationId;
  readonly title: string;
  readonly modelHint: string | null;
  readonly createdAt: string;
  readonly lastMessageAt: string;
  readonly archivedSummary: string | null;
  readonly messages: readonly ConversationMessage[];
}

export interface ConversationMessage {
  readonly id: string;
  readonly role: MessageRole;
  readonly content: string;
  readonly citations: readonly Citation[];
  readonly provider: AIProviderId | null;
  readonly model: string | null;
  readonly promptTokens: number;
  readonly completionTokens: number;
  readonly createdAt: string;
}

export interface Citation {
  readonly noteId: NoteId;
  readonly title: string;
  readonly chunkIndex: number;
  readonly rank: number;
}

/**
 * The domain facade. UI code (M12 / M13) talks to this. Implementations
 * compose `AIProvider` + the M10 vector store + the M04 notes service.
 */
export interface AIServices {
  // Per-note (one-shot)
  summarize(noteId: NoteId, length: SummaryLength): Promise<string>;
  suggestTags(noteId: NoteId): Promise<readonly string[]>;
  suggestTitle(noteId: NoteId): Promise<string>;
  extractActionItems(noteId: NoteId): Promise<readonly string[]>;
  rewrite(text: string, instruction: string): Promise<string>;

  // Corpus-wide (the second-brain features)
  chat(
    conversationId: ConversationId | null,
    message: string,
    scope?: NoteScope,
  ): AsyncIterable<ChatChunk>;
  relatedNotes(noteId: NoteId, k?: number): Promise<readonly RelatedNote[]>;
  dailyBriefing(date: string): Promise<Briefing>;

  // Conversation persistence
  listConversations(): Promise<readonly Conversation[]>;
  getConversation(id: ConversationId): Promise<Conversation>;
  newConversation(seed?: string): Promise<ConversationId>;
  deleteConversation(id: ConversationId): Promise<void>;
}

File: packages/android/core-domain/src/main/kotlin/app/notes/domain/AI.kt
package app.notes.domain

/**
 * Provider-agnostic AI value types — Kotlin twin of `packages/domain/src/ai.ts`.
 * The actual `AIProvider` / `AIServices` interfaces live in the
 * Android AI module because they use `Flow` (the same boundary we
 * keep for sync-core and attachments-android).
 */

enum class AICapability { STREAMING, EMBEDDINGS, VISION, TOOL_USE }

data class ModelDescriptor(
    val id: String,
    val contextTokens: Int,
    val supports: Set<AICapability>,
) {
    init {
        require(id.isNotBlank()) { "ModelDescriptor.id cannot be empty" }
        require(contextTokens > 0) { "ModelDescriptor.contextTokens must be > 0" }
    }
}

enum class MessageRole(val wireName: String) {
    SYSTEM("system"),
    USER("user"),
    ASSISTANT("assistant"),
    ;

    companion object {
        fun fromWire(name: String): MessageRole? = entries.firstOrNull { it.wireName == name }
    }
}

data class AIMessage(val role: MessageRole, val content: String) {
    init { require(content.isNotEmpty()) { "AIMessage.content cannot be empty" } }
}

data class AIOptions(
    val model: String,
    val temperature: Double? = null,
    val maxTokens: Int? = null,
    val stop: List<String> = emptyList(),
) {
    init {
        require(model.isNotBlank()) { "AIOptions.model cannot be empty" }
        if (temperature != null) require(temperature in 0.0..2.0) { "temperature must be in [0, 2]" }
        if (maxTokens != null) require(maxTokens > 0) { "maxTokens must be > 0" }
    }
}

data class AIResponse(
    val content: String,
    val model: String,
    val promptTokens: Int,
    val completionTokens: Int,
    val latencyMs: Long,
)

data class AIChunk(
    val delta: String,
    val done: Boolean,
    val model: String? = null,
    val promptTokens: Int? = null,
    val completionTokens: Int? = null,
)

data class ProviderHealth(
    val ok: Boolean,
    val message: String? = null,
    val latencyMs: Long? = null,
)

enum class SummaryLength { SHORT, MEDIUM, LONG }

data class NoteScope(
    val notebookIds: List<NotebookId>? = null,
    val tagIds: List<TagId>? = null,
    val excludeAi: Boolean = true,
)

/**
 * Unified streaming chunk type for the AIServices.chat() flow. Tagged
 * union via [kind]; readers `when (chunk.kind)`-dispatch.
 */
data class ChatChunk(
    val kind: Kind,
    val delta: String? = null,
    val noteId: NoteId? = null,
    val title: String? = null,
    val chunkIndex: Int? = null,
    val rank: Int? = null,
    val conversationId: ConversationId? = null,
    val tokens: ChunkTokens? = null,
    val model: String? = null,
    val latencyMs: Long? = null,
    val warning: String? = null,
) {
    enum class Kind { DELTA, CITATION, DONE, WARNING }
}

data class ChunkTokens(val prompt: Int, val completion: Int)

data class RelatedNote(
    val noteId: NoteId,
    val title: String,
    val snippet: String,
    val distance: Double,
)

data class Briefing(
    val date: String,
    val markdown: String,
    val model: String,
    val tokens: ChunkTokens,
)

data class AIConversation(
    val id: ConversationId,
    val title: String,
    val modelHint: String?,
    val createdAt: String,
    val lastMessageAt: String,
    val archivedSummary: String?,
    val messages: List<ConversationMessage>,
)

data class ConversationMessage(
    val id: String,
    val role: MessageRole,
    val content: String,
    val citations: List<Citation>,
    val provider: AIProviderId?,
    val model: String?,
    val promptTokens: Int,
    val completionTokens: Int,
    val createdAt: String,
)

data class Citation(
    val noteId: NoteId,
    val title: String,
    val chunkIndex: Int,
    val rank: Int,
)
